src/ucs/datastruct/array.c

```diff
@@ -13,31 +13,51 @@
 #include <ucs/sys/math.h>
 #include <ucs/debug/log.h>
 #include <ucs/debug/memtrack_int.h>
+#include <ucs/debug/assert.h>
 
 
 /* Increase the array buffer length by this factor, whenever it needs to grow */
 #define UCS_ARRAY_GROW_FACTOR   2
 
 
-ucs_status_t ucs_array_grow(void **buffer_p, size_t *capacity_p,
-                            size_t min_capacity, size_t value_size,
-                            const char *array_name, const char *value_name)
+ucs_status_t
+ucs_array_grow(void **buffer_p, size_t *capacity_p, size_t min_capacity,
+               size_t value_size, ucs_array_free_callback_t free_cb,
+               void *free_cb_arg, const char *array_name,
```

Comment:
maybe simpler to return the released pointer instead of defining a callback?
if nothing was released - return NULL